clients/src/main/java/org/apache/kafka/clients/consumer/internals/MembershipManagerImpl.java

```diff
@@ -160,92 +228,91 @@ private boolean maybeTransitionToStable() {
 
+    /**
+     * {@inheritDoc}
+     */
     @Override
     public AssignorSelection assignorSelection() {
         return this.assignorSelection;
     }
 
+    /**
+     * {@inheritDoc}
+     */
     @Override
-    public ConsumerGroupHeartbeatResponseData.Assignment assignment() {
+    public ConsumerGroupHeartbeatResponseData.Assignment currentAssignment() {
         return this.currentAssignment;
     }
 
+
+    /**
+     * Assignment that the member received from the server but hasn't completely processed yet.
+     */
     // VisibleForTesting
     Optional<ConsumerGroupHeartbeatResponseData.Assignment> targetAssignment() {
         return targetAssignment;
     }
 
-    // VisibleForTesting
-    Optional<ConsumerGroupHeartbeatResponseData.Assignment> nextTargetAssignment() {
-        return nextTargetAssignment;
-    }
-
     /**
-     * Set the current assignment for the member. This indicates that the reconciliation of the
-     * target assignment has been successfully completed.
-     * This will clear the {@link #targetAssignment}, and take on the
-     * {@link #nextTargetAssignment} if any.
+     * This indicates that the reconciliation of the target assignment has been successfully
+     * completed, so it will make it effective by assigning it to the current assignment.
      *
-     * @param assignment Assignment that has been successfully processed as part of the
-     *                   reconciliation process.
+     * @params Assignment that has been successfully reconciled. This is expected to
+     * match the target assignment defined in {@link #targetAssignment()}
      */
     @Override
-    public void updateAssignment(ConsumerGroupHeartbeatResponseData.Assignment assignment) {
-        this.currentAssignment = assignment;
-        if (!nextTargetAssignment.isPresent()) {
-            targetAssignment = Optional.empty();
-        } else {
-            targetAssignment = Optional.of(nextTargetAssignment.get());
-            nextTargetAssignment = Optional.empty();
+    public void onTargetAssignmentProcessComplete(ConsumerGroupHeartbeatResponseData.Assignment assignment) {
+        if (assignment == null) {
+            throw new IllegalArgumentException("Assignment cannot be null");
         }
+        if (!assignment.equals(targetAssignment.orElse(null))) {
+            // This could be simplified to remove the assignment param and just assume that what
+            // was reconciled was the targetAssignment, but keeping it explicit and failing fast
+            // here to uncover any issues in the interaction of the assignment processing logic
+            // and this.
+            throw new IllegalStateException(String.format("Reconciled assignment %s does not " +
+                            "match the initial target assignment %s", assignment, targetAssignment.orElse(null)));
+        }
+        this.currentAssignment = assignment;
+        targetAssignment = Optional.empty();
         maybeTransitionToStable();
```

Comment:
Do we still need `maybeTransitionToStable`? It seems that we could just transition to Stable here as we don't have any pending assignments now.

I also wonder how we will signal to the heartbeat manager that it must heartbeat immediately to ack the assignment. How do you see this?

Reply:
Agree that we can directly transition to stable here. Done.

As for triggering the HB when needed (outside of interval), @kirktrue , @philipnee and I just synced and are considering to base it on the existing `shouldHeartbeatNow` as you pointed out, with the revised approach of having the reconciliation logic triggered from within the `membershipManager` transition to reconciling, and updating it when the transition completes. 

The [reconciler PR](https://github.com/apache/kafka/pull/14357) will propose the approach for better aligning the components and interactions, but this is what we agreed on:
 
- HB manager using the membershipManager to trigger transitions, and to check if it `shouldSendHeartbeatNow`, that's it.  
- MembershipManager triggering the reconciliation logic and updating the `shouldSendHeartbeatNow` accordingly when the reconciliation completes.
- Reconciler only responsible for determine what to reconcile (diff between subscription and new assignment) & invoke call backs

This is all part of the [reconciliation PR](https://github.com/apache/kafka/pull/14357) so we can follow-up the changes/conversation there.